### PR TITLE
Add crack tip enrichment for X-FEM fracture

### LINKED ISSUE #10042: Add near-tip enrichment functions to X-FEM
## Rationale
The original X-FEM represents the cracks with enriched shape functions consist of both Heaviside and near-tip asymptotic functions.The initial implementation of the X-FEM in MOOSE is based on the phantom node method which is equivalent to the original X-FEM formulation with Heaviside enrichment. With phantom node method, cracks must terminate along element faces. This introduces a somewhat artificial change to the crack front geometry. In addition, the lack of near-tip enrichment reduces the capability of capturing the singular nature of the crack front fields. Thus, it may be desirable or even necessary to incorporate the near-tip enrichment to improve the accuracy of fracture simulation.

## Description
screen shot 2017-10-09 at 9 40 52 am

## Impact
Add new stress divergence kernel and strain material class. Won't affect existing X-FEM implementation.

### PR DESCRIPTION
**Pull request by jiangwen84**
screen shot 2017-10-09 at 9 40 52 am

This PR adds the third part contribution to the displacement field. The displacement field u(x) now contains both the standard and the enrichment solution. A new `ComputeStrain` Material Class is implemented to calculate the strains from the overall displacements. Additional `StressDivergenceKernel` is also added to obtain the residual and jacobian contribution from those enrichment variables. The tasks of adding new variables, kernels and boundary conditions are incorporated into `XFEM` Action which greatly facilitates the use of near-tip enrichment.

Both 2D and 3D tests are added. 

close #10042

### CODE REVIEW

File: docs/content/documentation/systems/BCs/xfem/CrackTipEnrichmentCutOffBC.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
+# CrackTipEnrichmentCutOffBC
+MOOSE currently does not permit to add variables on a subset of the domain. An in-efficient workaround is to add new variables associated with the enrichment shape functions for the whole mesh and then to simply fix DOFs to zero for those nodes with basis function supports that are far away from any crack tip. This approach is implemented in MOOSE for simplicity and parameter `cut_off_radius` is used to specify the cut-off radius for the enrichment variables.
```
Comment (bwspenc) [inline docs/content/documentation/systems/BCs/xfem/CrackTipEnrichmentCutOffBC.md]: I'd reword the first couple sentences like this:

MOOSE permits variables to be restricted to specific blocks, but currently does not provide a straightforward way to add variables to just the enriched region near the crack tip. An inefficient workaround is to add new variables associated with the enrichment shape functions for the whole mesh and then to simply fix DOFs to zero for those nodes with basis function supports that are far away from any crack tip.

File: modules/xfem/test/tests/crack_tip_enrichement/penny_crack_3d.i [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+[]
+
+[Variables]
+  [./disp_x]
+  [../]
+  [./disp_y]
+  [../]
+  [./disp_z]
+  [../]
+  [./enrich1_x]
```
Comment (bwspenc) [inline modules/xfem/test/tests/crack_tip_enrichement/penny_crack_3d.i]: Why not use the Action to set all this up?

**Review by bwspenc — CHANGES_REQUESTED** [review]
The general architecture looks good.  Hopefully my suggested changes shouldn't be too hard to make.

File: modules/xfem/test/tests/crack_tip_enrichement/penny_crack_3d.i (hunk 2 above)
Comment (jiangwen84) [reply modules/xfem/test/tests/crack_tip_enrichement/penny_crack_3d.i]: Done

File: modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  // Remove the Eigen strain
+  for (auto es : _eigenstrains)
+    _mechanical_strain[_qp] -= (*es)[_qp];
+}
+
+void
+ComputeCrackTipEnrichmentSmallStrain::initQpStatefulProperties()
+{
+  _enrich_strain[_qp].zero();
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C]: Done.

File: modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+ComputeCrackTipEnrichmentSmallStrain::computeQpProperties()
+{
+  unsigned int crack_front_point_index =
+      _crack_front_definition->calculateRThetaToCrackFront(_q_point[_qp], _r, _theta);
+
+  if (MooseUtils::absoluteFuzzyEqual(_r, 0.0))
+    mooseError("ComputeCrackTipEnrichmentSmallStrain: the distance between a point and the crack "
+               "tip/front is zero.");
+
+  Real st = std::sin(_theta);
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C]: Done.

File: modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    _B(4),
+    _dBX(4),
+    _dBx(4)
+{
+  for (unsigned int i = 0; i < _enrich_variable.size(); ++i)
+    _enrich_variable[i].resize(_ndisp);
+
+  const std::vector<NonlinearVariableName> & nl_vnames =
+      getParam<std::vector<NonlinearVariableName>>("enrichment_displacements");
+  NonlinearSystem & nl = _fe_problem.getNonlinearSystem();
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C]: Done

File: modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    _crack_front_definition(&getUserObject<CrackFrontDefinition>("crack_front_definition")),
+    _B(4),
+    _dBX(4),
+    _dBx(4)
+{
+  for (unsigned int i = 0; i < _enrich_variable.size(); ++i)
+    _enrich_variable[i].resize(_ndisp);
+
+  const std::vector<NonlinearVariableName> & nl_vnames =
+      getParam<std::vector<NonlinearVariableName>>("enrichment_displacements");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/materials/ComputeCrackTipEnrichmentSmallStrain.C]: Done

File: modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+  for (unsigned int i = 0; i < _enrich_disp_var.size(); ++i)
+  {
+    if (jvar == _enrich_disp_var[i])
+    {
+      coupled_component = i / 4;
+      coupled_enrichment_component = i % 4;
+      active = true;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C]: Yes, I added them. It turns out much better convergence.

File: modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+Real
+CrackTipEnrichmentStressDivergenceTensors::computeQpResidual()
+{
+  prepareCrackTipEnrichementFunctionAtNode(_current_elem->get_node(_i), _BI);
+
+  unsigned int crack_front_point_index =
+      _crack_front_definition->calculateRThetaToCrackFront(_q_point[_qp], _r, _theta);
+
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C]: I made a new class EnrichmentFunctionCalculation which performs all the calculation the values and derivatives of those enrichment functions. And I let CrackTipEnrichmentStressDivergenceTensors and ComputeCrackTipEnrichmentSmallStrain derive from it to avoid code duplication.

File: modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+#include "CrackTipEnrichmentStressDivergenceTensors.h"
+#include "ElasticityTensorTools.h"
+
+template <>
+InputParameters
+validParams<CrackTipEnrichmentStressDivergenceTensors>()
+{
+  InputParameters params = validParams<ALEKernel>();
+  params.addClassDescription("Enrich stress divergence kernel for the Cartesian coordinate system");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C]: Done

File: modules/xfem/src/bcs/CrackTipEnrichementCutOffBC.C [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    _cut_off_radius(getParam<Real>("cut_off_radius")),
+    _crack_front_definition(&getUserObject<CrackFrontDefinition>("crack_front_definition"))
+{
+}
+
+bool
+CrackTipEnrichmentCutOffBC::shouldApply()
+{
+  Real r, theta;
+  _crack_front_definition->calculateRThetaToCrackFront((*_current_node), r, theta);
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/bcs/CrackTipEnrichementCutOffBC.C]: Yes, they are expensive and it needs optimization in the future.

File: modules/xfem/src/bcs/CrackTipEnrichementCutOffBC.C [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+/****************************************************************/
+
+#include "CrackTipEnrichmentCutOffBC.h"
+
+template <>
+InputParameters
+validParams<CrackTipEnrichmentCutOffBC>()
+{
+  InputParameters p = validParams<PresetBC>();
+  p.addParam<Real>("cut_off_radius", 0.1, "The cut off radius of crack tip enrichment functions");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/bcs/CrackTipEnrichementCutOffBC.C]: I cannot make it required as it is not needed for the case without crack tip enrichment. I get rid of the default though.

File: modules/xfem/src/actions/XFEMAction.C [hunk 12]
```diff
@@ -47,12 +57,21 @@ XFEMAction::XFEMAction(InputParameters params)
...
   _family = "MONOMIAL";
   if (isParamValid("output_cut_plane"))
     _xfem_cut_plane = getParam<bool>("output_cut_plane");
+
+  if (_use_crack_tip_enrichment)
+  {
+    _crack_front_definition = getParam<UserObjectName>("crack_front_definition");
+    _enrich_displacements = getParam<std::vector<VariableName>>("enrichment_displacements");
+    _cut_off_bc = getParam<std::vector<BoundaryName>>("cut_off_boundary");
+    _cut_off_radius = getParam<Real>("cut_off_radius");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/actions/XFEMAction.C]: Done

File: modules/xfem/src/actions/XFEMAction.C [hunk 13]
```diff
@@ -47,12 +57,21 @@ XFEMAction::XFEMAction(InputParameters params)
...
 {
   _order = "CONSTANT";
   _family = "MONOMIAL";
   if (isParamValid("output_cut_plane"))
     _xfem_cut_plane = getParam<bool>("output_cut_plane");
+
+  if (_use_crack_tip_enrichment)
+  {
+    _crack_front_definition = getParam<UserObjectName>("crack_front_definition");
+    _enrich_displacements = getParam<std::vector<VariableName>>("enrichment_displacements");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/actions/XFEMAction.C]: Per our discussion, we leave it as-it for new.

File: modules/xfem/src/actions/XFEMAction.C [hunk 14]
```diff
@@ -38,6 +39,15 @@ validParams<XFEMAction>()
...
   params.addParam<Real>("crack_growth_increment", 0.1, "Crack growth increment");
+  params.addParam<bool>("use_crack_tip_enrichment", false, "Use crack tip enrichment functions");
+  params.addParam<UserObjectName>("crack_front_definition",
+                                  "The CrackFrontDefinition user object name");
+  params.addParam<std::vector<VariableName>>("enrichment_displacements",
+                                             "String of enrichment displacement");
+  params.addParam<std::vector<BoundaryName>>(
+      "cut_off_boundary", "Select all nodes and fix the DOFs away from crack tip");
+  params.addParam<Real>(
+      "cut_off_radius", 0.1, "The cut off radius of crack tip enrichment functions");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/actions/XFEMAction.C]: Done

File: modules/xfem/src/actions/XFEMAction.C [hunk 15]
```diff
@@ -38,6 +39,15 @@ validParams<XFEMAction>()
   params.addParam<bool>("output_cut_plane", false, "Output the XFEM cut plane and volume fraction");
   params.addParam<bool>("use_crack_growth_increment", false, "Use fixed crack growth increment");
   params.addParam<Real>("crack_growth_increment", 0.1, "Crack growth increment");
+  params.addParam<bool>("use_crack_tip_enrichment", false, "Use crack tip enrichment functions");
+  params.addParam<UserObjectName>("crack_front_definition",
+                                  "The CrackFrontDefinition user object name");
+  params.addParam<std::vector<VariableName>>("enrichment_displacements",
+                                             "String of enrichment displacement");
+  params.addParam<std::vector<BoundaryName>>(
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/actions/XFEMAction.C]: Done

File: modules/xfem/src/actions/XFEMAction.C [hunk 16]
```diff
@@ -38,6 +39,15 @@ validParams<XFEMAction>()
   params.addParam<bool>("output_cut_plane", false, "Output the XFEM cut plane and volume fraction");
   params.addParam<bool>("use_crack_growth_increment", false, "Use fixed crack growth increment");
   params.addParam<Real>("crack_growth_increment", 0.1, "Crack growth increment");
+  params.addParam<bool>("use_crack_tip_enrichment", false, "Use crack tip enrichment functions");
+  params.addParam<UserObjectName>("crack_front_definition",
+                                  "The CrackFrontDefinition user object name");
+  params.addParam<std::vector<VariableName>>("enrichment_displacements",
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/actions/XFEMAction.C]: Done

File: modules/xfem/src/actions/XFEMAction.C [hunk 17]
```diff
@@ -38,6 +39,15 @@ validParams<XFEMAction>()
   params.addParam<bool>("output_cut_plane", false, "Output the XFEM cut plane and volume fraction");
   params.addParam<bool>("use_crack_growth_increment", false, "Use fixed crack growth increment");
   params.addParam<Real>("crack_growth_increment", 0.1, "Crack growth increment");
+  params.addParam<bool>("use_crack_tip_enrichment", false, "Use crack tip enrichment functions");
+  params.addParam<UserObjectName>("crack_front_definition",
+                                  "The CrackFrontDefinition user object name");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/src/actions/XFEMAction.C]: Done

File: modules/xfem/include/materials/ComputeCrackTipEnrichmentSmallStrain.h [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+#include "Material.h"
+#include "RankTwoTensor.h"
+#include "RankFourTensor.h"
+#include "RotationTensor.h"
+#include "ComputeStrainBase.h"
+#include "Assembly.h"
+#include "CrackFrontDefinition.h"
+
+/**
+ * ComputeCrackTipEnrichmentSmallStrain calcualte the sum of standard strain and enrichement strain
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/include/materials/ComputeCrackTipEnrichmentSmallStrain.h]: Done

File: modules/xfem/include/bcs/CrackTipEnrichmentCutOffBC.h [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+class CrackTipEnrichmentCutOffBC : public PresetBC
+{
+public:
+  CrackTipEnrichmentCutOffBC(const InputParameters & parameters);
+
+protected:
+  virtual bool shouldApply() override;
+
+  const Real _cut_off_radius;
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/xfem/include/bcs/CrackTipEnrichmentCutOffBC.h]: Done

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 20]
```diff
@@ -1208,6 +1216,123 @@ CrackFrontDefinition::calculateRThetaToCrackFront(const Point qp,
...
+    }
+  }
+
+  Point closest_point(0.0);
+  RealVectorValue closest_point_to_p;
+
+  const Point * crack_front_point = getCrackFrontPoint(point_index);
+  RealVectorValue crack_front_point_rot = rotateToCrackFrontCoords(*crack_front_point, point_index);
+
+  RealVectorValue crack_front_edge =
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: Done

File: modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h [hunk 21]
```diff
@@ -54,10 +54,13 @@ class CrackFrontDefinition : public GeneralUserObject, public BoundaryRestrictab
                                            const unsigned int point_index) const;
   RankTwoTensor rotateToCrackFrontCoords(const RankTwoTensor tensor,
                                          const unsigned int point_index) const;
+  RealVectorValue rotateFromCrackFrontCoordsToGlobal(const RealVectorValue vector,
+                                                     const unsigned int point_index) const;
   void calculateRThetaToCrackFront(const Point qp,
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h]: Done

File: modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h [hunk 22]
```diff
@@ -54,10 +54,13 @@ class CrackFrontDefinition : public GeneralUserObject, public BoundaryRestrictab
                                            const unsigned int point_index) const;
   RankTwoTensor rotateToCrackFrontCoords(const RankTwoTensor tensor,
                                          const unsigned int point_index) const;
+  RealVectorValue rotateFromCrackFrontCoordsToGlobal(const RealVectorValue vector,
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h]: Done

File: docs/content/documentation/systems/Materials/xfem/ComputeCrackTipEnrichmentSmallStrain.md [hunk 23]
```diff
@@ -0,0 +1,12 @@
+
+# ComputeCrackTipEnrichmentSmallStrain
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply docs/content/documentation/systems/Materials/xfem/ComputeCrackTipEnrichmentSmallStrain.md]: Done

File: docs/content/documentation/systems/Kernels/xfem/CrackTipEnrichmentStressDivergenceTensors.md [hunk 24]
```diff
@@ -0,0 +1,12 @@
+
+# CrackTipEnrichmentStressDivergenceTensors
+The displacement field contains both the standard and the enrichment solution. `CrackTipEnrichmentStressDivergenceTensors` is added to obtain the residual and jacobian contribution from those enrichment variables.
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply docs/content/documentation/systems/Kernels/xfem/CrackTipEnrichmentStressDivergenceTensors.md]: Done

File: docs/content/documentation/systems/BCs/xfem/CrackTipEnrichmentCutOffBC.md (hunk 1 above)
Comment (jiangwen84) [reply docs/content/documentation/systems/BCs/xfem/CrackTipEnrichmentCutOffBC.md]: Done

**Comment by jiangwen84** [comment]
@bwspenc I addressed all the comments:-)

**Comment by jiangwen84** [comment]
I made the cut_off_bc required param in BC and we can see if it is conflicted with the action setup.

File: modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h [hunk 25]
```diff
@@ -54,10 +54,25 @@ class CrackFrontDefinition : public GeneralUserObject, public BoundaryRestrictab
                                            const unsigned int point_index) const;
   RankTwoTensor rotateToCrackFrontCoords(const RankTwoTensor tensor,
                                          const unsigned int point_index) const;
+
+  /*** rotate a vector from crack front cartesian coordinate to global cartesian coordinate using
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h]: Sorry to be a pain here, but these need to be formatted like this:
```
/**
  * Text
  * @param param_a description of param_a
  */
```

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 26]
```diff
@@ -1208,6 +1216,32 @@ CrackFrontDefinition::calculateRThetaToCrackFront(const Point qp,
...
+      min_dist = dist;
+      point_index = pit;
+    }
+  }
+
+  calculateRThetaToCrackFront(qp, point_index, r, theta);
+
+  return point_index;
+}
+
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: Much better! 👍

File: modules/xfem/include/base/EnrichmentFunctionCalculation.h [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "CrackFrontDefinition.h"
+
+class EnrichmentFunctionCalculation;
+
+template <>
+InputParameters validParams<EnrichmentFunctionCalculation>();
+
+/**
+ * Performe calculation of enrichment function values and derivatives.
```
Comment (bwspenc) [inline modules/xfem/include/base/EnrichmentFunctionCalculation.h]: Performe -> Perform

File: modules/xfem/include/base/EnrichmentFunctionCalculation.h [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+ */
+class EnrichmentFunctionCalculation
+{
+public:
+  EnrichmentFunctionCalculation(const CrackFrontDefinition * crack_front_definition);
+
+  /*** calculate the enrichment function values at point
```
Comment (bwspenc) [inline modules/xfem/include/base/EnrichmentFunctionCalculation.h]: Make these match doxygen formatting as well

File: modules/xfem/include/kernels/CrackTipEnrichmentStressDivergenceTensors.h [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+public:
+  CrackTipEnrichmentStressDivergenceTensors(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual() override;
+  virtual Real computeQpJacobian() override;
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
+
+  std::string _base_name;
+
```
Comment (bwspenc) [inline modules/xfem/include/kernels/CrackTipEnrichmentStressDivergenceTensors.h]: Add doxygen comments for these variables

File: modules/xfem/include/materials/ComputeCrackTipEnrichmentSmallStrain.h [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+public:
+  ComputeCrackTipEnrichmentSmallStrain(const InputParameters & parameters);
+  virtual ~ComputeCrackTipEnrichmentSmallStrain() {}
+
+protected:
+  virtual void computeProperties() override;
+
+  virtual void computeQpProperties() override;
+
+  std::vector<Real> _enrich_disp;
```
Comment (bwspenc) [inline modules/xfem/include/materials/ComputeCrackTipEnrichmentSmallStrain.h]: Doxygen comments

**Review by bwspenc — CHANGES_REQUESTED** [review]
It's really close -- it just needs the documentation in the header files to be cleaned up a little bit to match doxygen formatting.

**Comment by jiangwen84** [comment]
@bwspenc comments are addressed:-)

**Review by bwspenc — APPROVED** [review]
